File: src/main/java/tn/finhub/TestLedger.java
package tn.finhub;

import tn.finhub.model.WalletTransaction;
import tn.finhub.model.TransactionManager;
import tn.finhub.model.WalletModel;
import tn.finhub.model.Wallet;

import java.util.List;
import java.math.BigDecimal;

public class TestLedger {
    public static void main(String[] args) {
        WalletModel walletModel = new WalletModel();
        TransactionManager txDAO = new TransactionManager();

        try {
            java.sql.Connection conn = walletModel.getConnection();
            java.sql.PreparedStatement ps = conn.prepareStatement("SELECT id FROM wallet LIMIT 5");
            java.sql.ResultSet rs = ps.executeQuery();
            while (rs.next()) {
                int walletId = rs.getInt(1);
                Wallet w = walletModel.findById(walletId);
                System.out.println("--- WALLET " + walletId + " ---");
                System.out.println("Balance: " + w.getBalance() + ", Escrow: " + w.getEscrowBalance());

                List<WalletTransaction> txs = txDAO.findByWalletId(walletId);
                BigDecimal calcBalance = BigDecimal.ZERO;
                BigDecimal calcEscrow = BigDecimal.ZERO;

                for (WalletTransaction tx : txs) {
                    BigDecimal amt = tx.getAmount();
                    System.out.println("  " + tx.getType() + " : " + amt);
                    switch (tx.getType()) {
                        case "CREDIT":
                        case "DEPOSIT":
                        case "RELEASE":
                        case "TRANSFER_RECEIVED":
                        case "GENESIS":
                        case "ESCROW_RCVD":
                        case "ESCROW_FEE":
                        case "ESCROW_REFUND":
                            calcBalance = calcBalance.add(amt);
                            break;
                        case "DEBIT":
                        case "HOLD":
                        case "TRANSFER_SENT":
                            calcBalance = calcBalance.subtract(amt);
                            break;
                    }
                    if ("HOLD".equals(tx.getType()))
                        calcEscrow = calcEscrow.add(amt);
                    if ("RELEASE".equals(tx.getType()) || "ESCROW_SENT".equals(tx.getType())
                            || "ESCROW_REFUND".equals(tx.getType()))
                        calcEscrow = calcEscrow.subtract(amt);
                }

                System.out.println("CALC BAL: " + calcBalance + " == ACTUAL " + w.getBalance());
                System.out.println("CALC ESCROW: " + calcEscrow + " == ACTUAL " + w.getEscrowBalance());
                System.out.println("verifyBalance returned: " + walletModel.verifyBalance(walletId));
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: test_ledger.py
import mysql.connector
import os
import decimal

def get_db_connection():
    return mysql.connector.connect(
        host=os.getenv('DB_HOST', 'pidev-mariadb'),
        user=os.getenv('DB_USER', 'pidev_user'),
        password=os.getenv('DB_PASSWORD', 'Pidev@1122SpidevS'),
        database=os.getenv('DB_NAME', 'finhub')
    )

def test_ledger():
    try:
        conn = get_db_connection()
        cursor = conn.cursor(dictionary=True)
        cursor.execute("SELECT * FROM wallet LIMIT 5;")
        wallet = cursor.fetchall()
        print("--- WALLETS ---")
        for w in wallet:
            print(f"ID: {w['id']} | Balance: {w['balance']} | Escrow: {w['escrow_balance']} | Status: {w['status']}")
            
            cursor.execute("SELECT id, type, amount, reference FROM wallet_transaction WHERE wallet_id = %s ORDER BY created_at ASC", (w['id'],))
            txs = cursor.fetchall()
            calc_balance = decimal.Decimal(0)
            calc_escrow = decimal.Decimal(0)
            print("  Transactions:")
            for tx in txs:
                print(f"    - {tx['type']} : {tx['amount']}")
                amt = tx['amount']
                if tx['type'] in ["CREDIT", "DEPOSIT", "RELEASE", "TRANSFER_RECEIVED", "GENESIS", "ESCROW_RCVD", "ESCROW_FEE", "ESCROW_REFUND"]:
                    calc_balance += amt
                elif tx['type'] in ["DEBIT", "HOLD", "TRANSFER_SENT"]:
                    calc_balance -= amt
                
                if tx['type'] == "HOLD":
                    calc_escrow += amt
                elif tx['type'] in ["RELEASE", "ESCROW_SENT", "ESCROW_REFUND"]:
                    calc_escrow -= amt

            print(f"  CALC BALANCE: {calc_balance} vs ACTUAL: {w['balance']}")
            print(f"  CALC ESCROW: {calc_escrow} vs ACTUAL: {w['escrow_balance']}")
            if calc_balance != w['balance'] or calc_escrow != w['escrow_balance']:
                print("  !!! MISMATCH DETECTED !!!")

    except Exception as e:
        print(f"Error: {e}")
    finally:
        if 'conn' in locals() and conn.is_connected():
            conn.close()

if __name__ == "__main__":
    test_ledger()
